### PR TITLE
[flang][OpenMP] Enable lastprivate with collapse

### PR DESCRIPTION
**Pull request by NimishMishra**
This PR enables the lowering of lastprivate in presence of collapse clause.

### CODE REVIEW

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 1]
```diff
@@ -143,7 +143,7 @@ void DataSharingProcessor::collectOmpObjectListSymbol(
 }
 
 void DataSharingProcessor::collectSymbolsForPrivatization() {
-  bool hasCollapse = false;
+
```
Comment (tblah) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:146]: super nit: an empty line at the start of a function looks wrong to me (feel free to ignore)

**Review by tblah — APPROVED** [review]
LGTM, thanks!

**Comment by kiranchandramohan** [comment]
Have you checked whether the condition generation for the last iteration is correct when collapse is specified?

**Review by luporl — COMMENTED** [review]
As @kiranchandramohan mentioned, `DataSharingProcessor::insertLastPrivateCompare` is currently assuming collapse == 1.

While this patch may produce correct results, it would result in the lastprivate variable being assigned to on every iteration of the inner loops, when on the last iteration of the outer loop, which doesn't seem right.

However, checking all related iteration variables during many iterations would not be very efficient. IIRC, some time ago it was suggested that lastprivate handling should be moved to `omp` dialect, to make it possible to update the lastprivate variable in the end of the loop and avoid iteration variable comparisons.

**Comment by kiranchandramohan** [comment]
> As @kiranchandramohan mentioned, `DataSharingProcessor::insertLastPrivateCompare` is currently assuming collapse == 1.
> 
> While this patch may produce correct results, it would result in the lastprivate variable being assigned to on every iteration of the inner loops, when on the last iteration of the outer loop, which doesn't seem right.
> 
> However, checking all related iteration variables during many iterations would not be very efficient. IIRC, some time ago it was suggested that lastprivate handling should be moved to `omp` dialect, to make it possible to update the lastprivate variable in the end of the loop and avoid iteration variable comparisons.

The previous suggestion was to create an `omp.lastprivate` operation that takes the place of the `fir.if` that we insert for the lastprivate comparison. Now that @ergawy has made progress with the delayed privatization, you might be able to use the infrastructure he created to do this. The idea is to use the `%.omp.is_last` variable (which identifies whether this is the last iteration or not) that is given as input to the `__kmpc_for_static_init_4` openmp runtime call. You can use this in the exit block of the loop and do a comparison and branch to either do the lastprivate update (if it is the last iteration) or exit. All this to be done during translation or in the OpenMP IRBuilder.

```
call void @__kmpc_for_static_init_4(ptr @1, i32 %2, i32 34, ptr %.omp.is_last, ptr %.omp.lb, ptr %.omp.ub, ptr %.omp.stride, i32 1, i32 1)
; ... intervening code
omp.loop.exit:                                    ; preds = %omp.inner.for.end
  call void @__kmpc_for_static_fini(ptr @1, i32 %2)
  %14 = load i32, ptr %.omp.is_last, align 4
  %15 = icmp ne i32 %14, 0
  br i1 %15, label %.omp.lastprivate.then, label %.omp.lastprivate.done

.omp.lastprivate.then:                            ; preds = %omp.loop.exit
  %16 = load i32, ptr %x2, align 4
  store i32 %16, ptr %0, align 4
  br label %.omp.lastprivate.done

.omp.lastprivate.done:                            ; preds = %.omp.lastprivate.then, %omp.loop.exit
  ret void
  ```

File: flang/test/Lower/OpenMP/wsloop-collapse.f90 [hunk 2]
```diff
@@ -94,3 +94,36 @@ program wsloop_collapse
...
+               do k = 1, 30
+                 x = x + 1
+               end do
+             end do
+           end do
+!CHECK: fir.if {{.*}} {
+!CHECK: %[[LOADED_X:.*]] = fir.load %[[INNER_X_DECLARE]]#0 : !fir.ref<i32>
+!CHECK: hlfir.assign %[[LOADED_X]] to %[[VAL_X_DECLARE]]#0 temporary_lhs : i32, !fir.ref<i32>
+!CHECK: }
+        !$omp end parallel do
```
Comment (ergawy) [inline flang/test/Lower/OpenMP/wsloop-collapse.f90:128]: nit: similar to https://github.com/llvm/llvm-project/pull/93786#discussion_r1630939305, can you add the terminators to make the region structure more obvious?

**Review by ergawy — APPROVED** [review]
LG, just one nit.

**Comment by DavidTruby** [comment]
Hi, just removing the current check doesn't quite do the right thing, as the value of the lastprivate variable will be a random one of the last values of the inner loops, rather than the definite final one. To implement this we need to check that we have reached the final value of _all_ the loop indices, not just the outer one.

I've submitted a patch with this approach here: #99500

**Comment by NimishMishra** [comment]
> Hi, just removing the current check doesn't quite do the right thing, as the value of the lastprivate variable will be a random one of the last values of the inner loops, rather than the definite final one. To implement this we need to check that we have reached the final value of _all_ the loop indices, not just the outer one.
> 
> I've submitted a patch with this approach here: #99500

Hi @DavidTruby 

Thanks for the PR. I see what you mean. I'll close this then.

I've an internal reprioritization for a few days; I do plan to return to working on this and implementing the delayed privatization approach, as comments here (and your PR) mention. But in the meantime, we can go ahead with your PR.

**Comment by NimishMishra** [comment]
Closing as the issue is covered by https://github.com/llvm/llvm-project/pull/99500

**Pull request closed without merging**